### PR TITLE
handling unsupported content as per issues/91

### PR DESCRIPTION
**Pull request by mohamed--abdel-maksoud**
ref: https://github.com/scrapinghub/splash/issues/91

### CODE REVIEW

**Comment by mohamed--abdel-maksoud** [comment]
Hello, I'd appreciate some help with an issue here:
on unsupported_content_handler (line 129 in browser_tab.py), I try to read the whole response and write it into a file. reply.readAll() doesn't seem to return the whole file, so the download is always corrupt ... any idea what I do wrong?

Thanks!

File: splash/browser_tab.py [hunk 1]
```diff
@@ -95,6 +100,53 @@ def _init_webpage(self, verbosity, network_manager, splash_proxy_factory,
         self.web_page.setNetworkAccessManager(network_manager)
         self.web_page.splash_proxy_factory = splash_proxy_factory
         self.web_page.render_options = render_options
+        
+        
+        def unsupported_content_handler(reply):
```
Comment (kmike) [inline splash/browser_tab.py]: could you please make it a method? You can pass different arguments using e.g. functools.partial

**Comment by kmike** [comment]
reply.readAll() likely doesn't work because it can be called only once, and qt may also need to call it to read the response. There is a workaround - keep track of chunks which are received and build a body from there, see e.g. https://github.com/scrapinghub/splash/pull/494/ implementation.

File: splash/browser_tab.py [hunk 2]
```diff
@@ -129,10 +139,15 @@ def unsupported_content_handler(self, reply):
             
             fname = os.path.join(self.download_directory, fname)
             
-            with open(fname, 'wb+') as f: f.write(content)
-            
-            self.web_view.setHtml("<html><head /><body><h1>file was downloaded from %s to %s</h1></body>"%(url, fname))
+            # NOTE trying to retrieve the already downloaded chunks, cannot find the request in _response_bodies
+            #req_id = self.http_client.network_manager._get_request_id(reply.request())
+            #bodies = self.http_client.network_manager._response_bodies[req_id]
+            #log.msg('request id: %d, %d chunks' %(req_id, len(bodies)))
```
Comment (mohamed--abdel-maksoud) [inline splash/browser_tab.py]: I tried to retrieve the already downloaded chunks, but cannot find them in response_bodies

**Comment by mohamed--abdel-maksoud** [comment]
Thanks for the tips! I still cannot download unsupported files in an optimal way, please see the comment at browser_tab.py:142

File: splash/browser_tab.py (hunk 2 above)
Comment (kmike) [inline splash/browser_tab.py]: is response_body_enable set to True? See http://splash.readthedocs.io/en/stable/scripting-ref.html#splash-response-body-enabled and http://splash.readthedocs.io/en/stable/api.html#arg-response-body

Comment (mohamed--abdel-maksoud) [inline splash/browser_tab.py]: Thanks for the tip!
I enabled the property in defaults.py then wrote the request's response body to a file as below, it still writes a part of the file (just as it did with reply.readAll). I scanned the code to see what might be wrong, didn't find any hints.
`
# defaults.py
#
RESPONSE_BODY_ENABLED = True

# browser_tab.py
#
            req_id = self.http_client.network_manager._get_request_id(reply.request())
            bodies = self.http_client.network_manager._response_bodies[req_id]
            log.msg('request id: %d, %d chunks %s' %(req_id, len(bodies), type(bodies[0]) ))
            
            with open(fname, 'w+b') as f: 
                f.write(bytearray(bodies))
`

**Comment by wpxgit** [comment]
Hi,

i've thought about it. Is it possible that that is the standard behavior of qt5. If i think about my browser he tries to get the file, if it realise that its a file for download it opens download manager and begins to download that file.
I've read a little bit in qt5 discussions and it seems that the solution for an unsupported file is to download download it with a separate QNetworkAccessManager. 

And it makes a little sense - the browser needs the first lines of the file to recognize if its an html file or an image, gzipped or else...
?

**Comment by mohamed--abdel-maksoud** [comment]
Hi @wpxgit ,
Indeed, great insight. I just tried the code on pdfs of several sizes and the unsupported type handler is called after downloading ~1.2 KB of the file, regardless of the whole file size.
Can we then conclude that issuing an http_get on the unsupported file (i.e. the last commit) carries very little overhead? would this approach be acceptable, @kmike ?

**Comment by wpxgit** [comment]
any chance to get this pull request merged?

**Pull request closed without merging**